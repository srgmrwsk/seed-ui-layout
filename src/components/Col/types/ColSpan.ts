export type ColSpan = boolean | number | 'auto';
