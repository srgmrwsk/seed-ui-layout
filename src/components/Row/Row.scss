.seed-row {
  display: flex;
  flex-wrap: wrap;
  list-style: none;
  padding: 0;
  margin: 0;

  @each $breakpoint in map-keys($grid-breakpoints) {
    @include bp-up($breakpoint) {
      @if $breakpoint == xs {
        &_gutter {
          > * {
            margin-top: var(--seed-row-gutter-y);
            padding-left: calc(var(--seed-row-gutter-x) / 2);
            padding-right: calc(var(--seed-row-gutter-x) / 2);
          }

          margin-top: calc(var(--seed-row-gutter-y) * -1);
          margin-left: calc(var(--seed-row-gutter-x) / -2);
          margin-right: calc(var(--seed-row-gutter-x) / -2);
        }
      } @else {
        &_gutter-#{$breakpoint} {
          > * {
            margin-top: var(--row-gutter-y-#{$breakpoint});
            padding-left: calc(var(--row-gutter-x-#{$breakpoint}) / 2);
            padding-right: calc(var(--row-gutter-x-#{$breakpoint}) / 2);
          }

          margin-top: calc(var(--row-gutter-y-#{$breakpoint}) * -1);
          margin-left: calc(var(--row-gutter-x-#{$breakpoint}) / -2);
          margin-right: calc(var(--row-gutter-x-#{$breakpoint}) / -2);
        }
      }
    }
  }
}
