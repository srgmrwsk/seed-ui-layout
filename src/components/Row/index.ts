export * from './types/RowGutter';
export * from './Row';

export { default as Row } from './Row';
