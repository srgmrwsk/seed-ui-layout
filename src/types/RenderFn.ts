import React from 'react';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type RenderFn = (component: React.ReactElement, ...args: []) => any;
