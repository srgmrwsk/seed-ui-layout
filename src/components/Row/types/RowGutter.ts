import { Sizing } from '../../../types/Sizing';

export type RowGutter = Sizing | [Sizing, Sizing];
