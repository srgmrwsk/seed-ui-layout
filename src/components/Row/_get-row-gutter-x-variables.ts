import { Maybe } from '../../types/Maybe';
import { formatSpacing } from '../../utils/get-spacing-variables';
import { getVariablesBy } from '../../utils/get-variables-by';

import { RowGutter } from './types/RowGutter';

function formatRowGutterX(v: Maybe<RowGutter>): Maybe<string> {
  return Array.isArray(v) ? formatSpacing(v[1]) : formatSpacing(v);
}

export const getRowGutterXVariables = getVariablesBy(formatRowGutterX);
