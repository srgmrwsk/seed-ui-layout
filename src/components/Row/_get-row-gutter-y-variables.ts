import { Maybe } from '../../types/Maybe';
import { formatSpacing } from '../../utils/get-spacing-variables';
import { getVariablesBy } from '../../utils/get-variables-by';

import { RowGutter } from './types/RowGutter';

function formatRowGutterY(v: Maybe<RowGutter>): Maybe<string> {
  return Array.isArray(v) ? formatSpacing(v[0]) : formatSpacing(v);
}

export const getRowGutterYVariables = getVariablesBy(formatRowGutterY);
