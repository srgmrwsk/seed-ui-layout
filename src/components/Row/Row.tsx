import React from 'react';
import cx from 'classnames';

import { ResponsiveValue } from '../../types/ResponsiveValue';
import { Align } from '../../types/Align';
import { Justify } from '../../types/Justify';
import { getStringStyles } from '../../utils/get-string-styles';
import { getBooleanStyles } from '../../utils/get-boolean-styles';

import { RowGutter } from './types/RowGutter';
import { getRowGutterYVariables } from './_get-row-gutter-y-variables';
import { getRowGutterXVariables } from './_get-row-gutter-x-variables';

export interface RowProps
  extends Omit<React.AllHTMLAttributes<HTMLElement>, 'as'> {
  as?: React.ElementType;
  align?: ResponsiveValue<Align>;
  justify?: ResponsiveValue<Justify>;
  gutter?: ResponsiveValue<RowGutter>;
}

function Row(
  {
    as: As = 'div',
    align,
    gutter,
    justify,
    className,
    style,
    children,
    ...elementProps
  }: RowProps,
  ref: React.Ref<HTMLElement>,
): JSX.Element {
  return (
    <As
      {...elementProps}
      ref={ref}
      className={cx(
        'seed-row',
        getBooleanStyles('seed-row_gutter', gutter),
        getStringStyles('_seed-align', align),
        getStringStyles('_seed-justify', justify),
      )}
      style={{
        ...getRowGutterYVariables('seed-row-gutter-y', gutter),
        ...getRowGutterXVariables('seed-row-gutter-x', gutter),
        ...style,
      }}
    >
      {children}
    </As>
  );
}

Row.displayName = 'Row';

export default React.forwardRef(Row);
