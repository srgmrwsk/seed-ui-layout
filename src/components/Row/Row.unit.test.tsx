import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';

import { RenderFn } from '../../types/RenderFn';
import { axe, create } from '../../utils/test-utils';

import Row, { RowProps } from './Row';

describe('Row', () => {
  const setup = (renderFn: RenderFn, props?: RowProps) =>
    renderFn(<Row {...props} />);

  it('should render correctly with default props', () => {
    const actual = setup(create);
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with as prop', () => {
    const actual = setup(create, { as: 'section' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with children prop', () => {
    const actual = setup(create, { children: <p>Element with text.</p> });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with align prop', () => {
    const actual = setup(create, { align: 'center' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with gutter prop', () => {
    const actual = setup(create, { gutter: 2 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with gutter array prop', () => {
    const actual = setup(create, { gutter: [2, 4] });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with justify prop', () => {
    const actual = setup(create, { justify: 'center' });
    expect(actual).toMatchSnapshot();
  });

  it('should meet accessibility guidelines', async () => {
    const wrapper = setup(renderToStaticMarkup);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
