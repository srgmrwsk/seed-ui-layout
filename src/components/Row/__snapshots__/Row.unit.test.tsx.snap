// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row should render correctly with align prop 1`] = `
<div
  class="seed-row _seed-align_center"
/>
`;

exports[`Row should render correctly with as prop 1`] = `
<section
  class="seed-row"
/>
`;

exports[`Row should render correctly with children prop 1`] = `
<div
  class="seed-row"
>
  <p>
    Element with text.
  </p>
</div>
`;

exports[`Row should render correctly with default props 1`] = `
<div
  class="seed-row"
/>
`;

exports[`Row should render correctly with gutter array prop 1`] = `
<div
  class="seed-row seed-row_gutter"
  style="--seed-row-gutter-y: 16px; --seed-row-gutter-x: 32px;"
/>
`;

exports[`Row should render correctly with gutter prop 1`] = `
<div
  class="seed-row seed-row_gutter"
  style="--seed-row-gutter-y: 16px; --seed-row-gutter-x: 16px;"
/>
`;

exports[`Row should render correctly with justify prop 1`] = `
<div
  class="seed-row _seed-justify_center"
/>
`;
